File: src/com/company/Scanner.java
package com.company;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import com.company.Token.*;
import static com.company.Token.TokenType.*;

public class Scanner {

    private final String source;
    private final List<Token> tokens = new ArrayList<>();
    private int start = 0;
    private int current = 0;
    private int line = 1;

    private static final Map<String, TokenType> reservedKeywordsMap;

    static {
        reservedKeywordsMap = new HashMap<>();
        reservedKeywordsMap.put("and", AND);
        reservedKeywordsMap.put("class",  CLASS);
        reservedKeywordsMap.put("else",   ELSE);
        reservedKeywordsMap.put("false",  FALSE);
        reservedKeywordsMap.put("for",    FOR);
        reservedKeywordsMap.put("fun",    FUN);
        reservedKeywordsMap.put("if",     IF);
        reservedKeywordsMap.put("nil",    NIL);
        reservedKeywordsMap.put("or",     OR);
        reservedKeywordsMap.put("print",  PRINT);
        reservedKeywordsMap.put("return", RETURN);
        reservedKeywordsMap.put("super",  SUPER);
        reservedKeywordsMap.put("this",   THIS);
        reservedKeywordsMap.put("true",   TRUE);
        reservedKeywordsMap.put("var",    VAR);
        reservedKeywordsMap.put("while",  WHILE);
    }


    public Scanner(String source) {
        this.source = source;
    }

    public List<Token> scanTokens() {

        while (!isAtEnd()){
            //Have we reached the end of a line we are scanning ??
            start = current;
            scanToken();
        }

        tokens.add(new Token(EOF, "", null, line));
        return tokens;
    }

    private void scanToken() {
        char c = advance();
        switch (c){
            case '(': addToken(LEFT_PAREN); break;
            case ')': addToken(RIGHT_PAREN); break;
            case '{': addToken(LEFT_BRACE); break;
            case '}': addToken(RIGHT_BRACE); break;
            case ',': addToken(COMMA); break;
            case '.': addToken(DOT); break;
            case '-': addToken(MINUS); break;
            case '+': addToken(PLUS); break;
            case ';': addToken(SEMICOLON); break;
            case '*': addToken(STAR); break;
            case '!':
                addToken(match('=') ? BANG_EQUAL : BANG);
                break;
            case '=':
                addToken(match('=') ? EQUAL_EQUAL : EQUAL);
                break;
            case '<' :
                addToken(match('=') ? LESS_EQUAL : LESS);
                break;
            case '>' :
                addToken(match('=') ? GREATER_EQUAL : GREATER);
                break;
            case '/':
                if (match('/')){
                    while (peek() != '\n' && !isAtEnd()) advance();
                }else{
                    addToken(SLASH);
                }
                break;
            case ' ':
            case '\r':
            case '\t':
                break;
            case '\n':
                line++;
                break;
            case '"':
                string();
                break;
            default:
                if(isDigit(c)){
                    number();
                }else if (isAlpha(c)) {
                    // Identify Keywords and reserved words like "var", "nil", "Boolean"
                    identifier();
                }else {
                    Latch.error(line, "Unexpected error");
                }
                break;
        }
    }

    private void identifier() {
        while (isAlphaNumeric(peek())) advance();

        String text = source.substring(start, current);
        TokenType tokenType = reservedKeywordsMap.get(text);

        if (tokenType == null) tokenType = IDENTIFIER;

        addToken(tokenType);
    }

    private boolean isAlphaNumeric(char peek) {
        return isAlpha(peek) || isDigit(peek);
    }

    private boolean isAlpha(char c) {
        return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c == '_';
    }

    // NUMBER LITERALS
    private void number() {
        while (isDigit(peek())) advance();

        //look for fractional number(part) !!
        if(peek() == '.' && isDigit(peekNext())){
            //consume the '.'
            advance();

            // to get the remaining decimal number
            while(isDigit(peek())) advance();
        }

        addToken(NUMBER, Double.parseDouble(source.substring(start, current)));
    }

    // Usually check the next token
    private char peekNext() {
        if (current + 1 >= source.length()) return '\0';
        return source.charAt(current + 1);
    }

    private boolean isDigit(char c) {

        return c >= '0'  && c <=9;
    }


    //STRING LITERALS
    private void string() {
        while (peek() != '"' && !isAtEnd()){
            if (peek() == '\n') line++;
            advance();
        }

        if (isAtEnd()){
            Latch.error(line, "Unterminated string");
            return;
        }



        // The closing "
        advance();

        String value = source.substring(start + 1, current - 1);
        addToken(STRING, value);
    }

    private char peek() {
        if (isAtEnd()) return '\0';
        return source.charAt(current);
    }

    private boolean match(char expected) {
        if (isAtEnd()) return false;
        if (source.charAt(current) != expected) return false;

        current++;
        return true;
    }

    private void addToken(Token.TokenType type) {
        addToken(type, null);
    }

    private void addToken(Token.TokenType type, Object literal){
        String text = source.substring(start, current);
        tokens.add(new Token(type, text, literal, line));
    }

    private char advance() {
        return source.charAt(current++);
    }

    private boolean isAtEnd() {
        return current >= source.length();
    }
}

File: components/scanner.go
package components

import (
	"scoop/semantics"
	"strconv"
)

type Scanner struct {
	source             string
	tokens             []semantics.Token
	start              int
	current            int
	line               int
	reservedKeyWordMap map[string]semantics.TokenType
}

func InitScanner(source string) *Scanner {
	keywords := map[string]semantics.TokenType{
		"and":    semantics.AND,
		"class":  semantics.CLASS,
		"else":   semantics.ELSE,
		"false":  semantics.FALSE,
		"for":    semantics.FOR,
		"fun":    semantics.FUN,
		"if":     semantics.IF,
		"nil":    semantics.NIL,
		"or":     semantics.OR,
		"print":  semantics.PRINT,
		"return": semantics.RETURN,
		"super":  semantics.SUPER,
		"this":   semantics.THIS,
		"true":   semantics.TRUE,
		"var":    semantics.VAR,
		"while":  semantics.WHILE,
	}
	return &Scanner{
		source:             source,
		start:              0,
		current:            0,
		line:               1,
		reservedKeyWordMap: keywords,
	}
}

func (s *Scanner) ScanTokens() []semantics.Token {
	for !s.isAtEnd() {
		s.start = s.current
		s.scanToken()
	}

	s.tokens = append(s.tokens, semantics.Token{
		TokenType: semantics.EOF,
		Lexeme:    "",
		Literal:   nil,
		Line:      s.line,
	})
	return s.tokens
}

func (s *Scanner) isAtEnd() bool {
	return s.current >= len(s.source)
}

func (s *Scanner) advance() byte {
	currentChar := s.source[s.current]
	s.current++
	return currentChar
}

func (s *Scanner) addEmptyToken(tokenType semantics.TokenType) {
	s.addToken(tokenType, nil)
}

func (s *Scanner) addToken(tokenType semantics.TokenType, literal interface{}) {
	text := s.source[s.start:s.current]
	s.tokens = append(s.tokens, semantics.Token{TokenType: tokenType, Lexeme: text, Literal: literal, Line: s.line})
}

func (s *Scanner) peek() byte {
	if s.isAtEnd() {
		return 0
	}
	return s.source[s.current]
}

func (s *Scanner) peekNext() byte {
	if s.current+1 >= len(s.source) {
		return 0
	}
	return s.source[s.current+1]
}

func (s *Scanner) scanToken() {
	character := rune(s.advance())
	switch character {
	case '(':
		s.addEmptyToken(semantics.LEFT_PAREN)
	case ')':
		s.addEmptyToken(semantics.RIGHT_PAREN)
	case '{':
		s.addEmptyToken(semantics.LEFT_BRACE)
	case '}':
		s.addEmptyToken(semantics.RIGHT_BRACE)
	case ',':
		s.addEmptyToken(semantics.COMMA)
	case '.':
		s.addEmptyToken(semantics.DOT)
	case '-':
		s.addEmptyToken(semantics.MINUS)
	case '+':
		s.addEmptyToken(semantics.PLUS)
	case ';':
		s.addEmptyToken(semantics.SEMICOLON)
	case '*':
		s.addEmptyToken(semantics.STAR)
	case '!':
		if s.match('=') {
			s.addEmptyToken(semantics.BANG_EQUAL)
		} else {
			s.addEmptyToken(semantics.BANG)
		}
	case '=':
		if s.match('=') {
			s.addEmptyToken(semantics.EQUAL_EQUAL)
		} else {
			s.addEmptyToken(semantics.EQUAL)
		}
	case '<':
		if s.match('=') {
			s.addEmptyToken(semantics.LESS_EQUAL)
		} else {
			s.addEmptyToken(semantics.LESS)
		}
	case '>':
		if s.match('=') {
			s.addEmptyToken(semantics.GREATER_EQUAL)
		} else {
			s.addEmptyToken(semantics.GREATER)
		}
	case '/':
		if s.match('/') {
			for s.peek() != '\n' && !s.isAtEnd() {
				s.advance()
			}
		} else {
			s.addEmptyToken(semantics.SLASH)
		}
	case ' ', '\r', '\t':
	case '\n':
		s.line++
	case '"':
		s.string()
	default:
		if s.isDigit(character) {
			s.number()
		} else if s.isAlpha(character) {
			s.identifier()
		} else {
			// suppose to throw error here from scoop
		}

	}

}

func (s *Scanner) string() {
	for s.peek() != '"' && !s.isAtEnd() {
		if s.peek() == '\n' {
			s.line++
		}
		s.advance()
	}

	if s.isAtEnd() {
		// throw error here again ==> do this after refactor
		return
	}

	s.advance()
	value := s.source[s.start+1 : s.current-1]
	s.addToken(semantics.STRING, value)
}

func (s *Scanner) number() {
	for s.isDigit(rune(s.peek())) {
		s.advance()
	}

	if s.peek() == '.' && s.isDigit(rune(s.peekNext())) {

		s.advance()

		for s.isDigit(rune(s.peek())) {
			s.advance()
		}
	}
	number, err := strconv.ParseFloat(s.source[s.start:s.current], 64)
	if err != nil {
		// do something here
	}
	s.addToken(semantics.NUMBER, number)
}

func (s *Scanner) identifier() {
	for s.isAlphanumeric(rune(s.peek())) {
		s.advance()
	}

	text := s.source[s.start:s.current]
	tokenType, found := s.reservedKeyWordMap[text]

	if !found {
		tokenType = semantics.IDENTIFIER
	}

	s.addEmptyToken(tokenType)
}

func (s *Scanner) isDigit(char rune) bool {
	if char >= '0' && char <= '9' {
		return true
	}
	return false
}

func (s *Scanner) isAlphanumeric(peek rune) bool {
	return s.isAlpha(peek) || s.isDigit(peek)
}

func (s *Scanner) isAlpha(char rune) bool {
	if char >= 'a' && char <= 'z' || char >= 'A' && char <= 'Z' || char == '_' {
		return true
	}
	return false
}

func (s *Scanner) match(expected byte) bool {
	if s.isAtEnd() {
		return false
	}
	if s.source[s.current] != expected {
		return false
	}
	s.current++
	return true
}
